rfc/plugins.md

```diff
@@ -0,0 +1,206 @@
+ember-cli-deploy. For example:
+
+```javascript
+module.exports = {
+  name: 'ember-cli-deploy-example-plugin',
+  createDeployPlugin: function(options){
+    return {
+      name: options.name,
+      willDeploy: function:(deployment){
+        // do something during the willDeploy phase of the pipeline
+      },
+      didDeploy: function:(deployment){
+        // do something during the didDeploy phase of the pipeline
+      },
+      // etc, see hooks section for a complete list of methods that
+      // may be implemented by a plugin
+    }
+  }
+}
+```
+
+This approach limits the risk of name conflicts at the top-level of the addon.
+It also allows for the plugin author to consult the addon instance during
+creation of the plugin object to make any contextual decisions necessary.
+Finally, it is agnostic with respect to the type of object the plugin is.
+It may be a POJO, a subclass of CoreObject, or maybe an ES6 class.
+
+The `options` argument passed to `createDeployPlugin` will have a `name`
+property. Usually, the `name` will be the plugin name sans the `ember-cli-deploy-`
+prefix, unless a name has been specified as described in Advanced Plugin
+Configuration below.
+
+### The `deployment` object
+
+For each high-level ember-cli-deploy operation, a `deployment` object is created.
+This object is passed to each hook that is invoked on the plugins. It has a number
+of properties that may be of use to a plugin:
+
+Property | file | info
+--- | --- | ---
+`ui` | - | The ember-cli UI object that can be used to write to stdout.
+`appConfig` | `config/environment.js` | The application configuration.
+`config` | stored in `config/deploy.js` | The configuration portion of `config/deploy.js` for the active environment
+`data` | - | Runtime information about the current operation. Plugins can set properties on this object for later use by themselves or another plugin.
+
+### Async operations in hooks
+
+Hook functions can return a promise to block the deployment pipeline.
+Since most deployment involves some sort of IO it makes senses that most
+plugins will want an async function to complete before continuing to the
+next step. If a plugin does not return a promise, then ember-cli-deploy
+proceeds immediately.
+
+If a promise from any of the plugins is rejected then the deployment
+pipeline will stop and ember-cli-deploy will exit. Returned promises that are
+rejected are treated as unrecoverable errors.
+
+### Configuration
+
+By convention, plugin configuration should be kept in `config/deploy.js` and scoped by
+the plugin's name. e.g. for an `ember-cli-deploy-example` plugin, the configuration would
+go in:
+
+```javascript
+// config/deploy.js
+module.exports = {
+  production: {
+    "example": {
+      bucket: "my-app-production",
+      awsAccessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      awsSecretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
+    }
+  }
+};
+```
```

Comment:
when you say "ordered list of tasks" are you proposing getting rid of the hooks and having each host app explicitly specify its own workflow, a la grunt? We thought of this but decided against it, because we want ember-cli-deploy to provide more structure than that. Plugins should execute on certain hooks by default, with the ability to override. Overriding should be a minority case, and would be handled in the config as specified below.

or am I misunderstanding your point?